src/serializer/functions.js

```diff
@@ -27,22 +28,38 @@ export class Functions {
     this.functions = functions;
     this.moduleTracer = moduleTracer;
     this.writeEffects = new Map();
-    this.nameToFunctionValue = new Map();
+    this.functionExpressions = new Map();
   }
 
   realm: Realm;
   functions: ?Array<string>;
+  // maps back from FunctionValue to the expression string
+  functionExpressions: Map<FunctionValue, string>;
   moduleTracer: ModuleTracer;
-  writeEffects: Map<string, Effects>;
-  nameToFunctionValue: Map<string, FunctionValue>;
+  writeEffects: Map<FunctionValue, Effects>;
 
   checkThatFunctionsAreIndependent() {
     let functions = this.functions;
-    invariant(functions, "This method should only be called if initialized with defined functions");
+    let recordedAdditionalFunctions: Map<FunctionValue, string> = new Map();
```

Comment:
High level, currently, we have two ways to get additional functions, and we have two separate chunks of code to parse, interpret and merge them into "calls" array.
I would put them into two private helper functions: _generateAdditionalFunctionCallsFromInput() and _generateAdditionalFunctionCallsFromDirective(). Both are called from checkThatFunctionsAreIndependent().

This will make the code self-documented and more readable.